src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionSelector.cs

```diff
@@ -18,36 +20,85 @@ namespace Microsoft.AspNetCore.Mvc.Internal
+        private static readonly ActionDescriptor[] EmptyActions = new ActionDescriptor[0];
+
+        private readonly IActionDescriptorCollectionProvider _actionDescriptorCollectionProvider;
         private readonly ActionConstraintCache _actionConstraintCache;
-        private ILogger _logger;
+        private readonly ILogger _logger;
+
+        private Cache _cache;
 
         /// <summary>
         /// Creates a new <see cref="ActionSelector"/>.
         /// </summary>
-        /// <param name="decisionTreeProvider">The <see cref="IActionSelectorDecisionTreeProvider"/>.</param>
+        /// <param name="actionDescriptorCollectionProvider">
+        /// The <see cref="IActionDescriptorCollectionProvider"/>.
+        /// </param>
         /// <param name="actionConstraintCache">The <see cref="ActionConstraintCache"/> that
         /// providers a set of <see cref="IActionConstraint"/> instances.</param>
         /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
         public ActionSelector(
-            IActionSelectorDecisionTreeProvider decisionTreeProvider,
+            IActionDescriptorCollectionProvider actionDescriptorCollectionProvider,
             ActionConstraintCache actionConstraintCache,
             ILoggerFactory loggerFactory)
         {
-            _decisionTreeProvider = decisionTreeProvider;
+            _actionDescriptorCollectionProvider = actionDescriptorCollectionProvider;
             _logger = loggerFactory.CreateLogger<ActionSelector>();
             _actionConstraintCache = actionConstraintCache;
         }
 
+        private Cache Current
+        {
+            get
+            {
+                var actions = _actionDescriptorCollectionProvider.ActionDescriptors;
+                var cache = Volatile.Read(ref _cache);
+
+                if (cache != null && cache.Version == actions.Version)
+                {
+                    return cache;
+                }
+
+                cache = new Cache(actions);
+                _cache = cache;
+                return cache;
+            }
+        }
+
         public IReadOnlyList<ActionDescriptor> SelectCandidates(RouteContext context)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var tree = _decisionTreeProvider.DecisionTree;
-            return tree.Select(context.RouteData.Values);
+            var cache = Current;
+
+            // The Cache works based on a string[] of the route values in a pre-calculated order. This code extracts
+            // those values in the correct order.
+            var keys = cache.RouteKeys;
+            var values = new string[keys.Length];
+            for (var i = 0; i < keys.Length; i++)
+            {
+                context.RouteData.Values.TryGetValue(keys[i], out object value);
+
+                if (value != null)
```

Comment:
I agree that `IDictionary<>` is a troll api. Should we put this on hold and sort that out?